### ISSUE TITLE
NIRCam coronagraph detector geometry

### ISSUE THREAD

**Issue by JarronL** [spacetelescope/webbpsf#350]
In NIRCam's `detector` setter attribute, it always passes the imaging SIAF aperture names to `DetectorGeometry`. However, if a Lyot stop is in the pupil wheel, the focal plane shifts on the detector surface. So, in order for the `pix2angle` function to work correctly, we should probably provide the appropriate aperture name

``` python
    @JWInstrument.detector.setter # override setter in this subclass
    def detector(self, value):
        """ Set detector, including reloading the relevant info from SIAF """
        if value.upper() not in self.detector_list:
            raise ValueError("Invalid detector. Valid detector names are: {}".format(', '.join(self.detector_list)))
        # set the channel based on the requested detector
        new_channel = 'long' if value[-1] == '5' else 'short'
        self._switch_channel(new_channel)
        self._detector = value.upper()

        self._detector_geom_info = DetectorGeometry(self.name, self._detectors[self._detector])
```

Perhaps change to something like:

```python
    @JWInstrument.detector.setter # override setter in this subclass
    def detector(self, value):
        """ Set detector, including reloading the relevant info from SIAF """
        if value.upper() not in self.detector_list:
            raise ValueError("Invalid detector. Valid detector names are: {}".format(', '.join(self.detector_list)))
        # set the channel based on the requested detector
        new_channel = 'long' if value[-1] == '5' else 'short'
        self._switch_channel(new_channel)
        self._detector = value.upper()

        # Detector aperture name
        # Need to send correct aperture name for coronagraphic masks due to detector shift
        if (self._image_mask is not None) and (self._image_mask in self._image_mask_apertures):
            apname = self._image_mask_apertures[self._image_mask]
        else:
            apname = self._detectors[self._detector]
        self._detector_geom_info = DetectorGeometry(self.name, apname)

```

Also, need to update `_get_fits_header`
```python
    def _get_fits_header(self, hdulist, options):
        """ Format NIRCam-like FITS headers, based on JWST DMS SRD 1 FITS keyword info """
        super(NIRCam, self)._get_fits_header(hdulist, options)

        hdulist[0].header['MODULE'] = (self.module, 'NIRCam module: A or B')
        hdulist[0].header['CHANNEL'] = ('Short' if self.channel == 'short' else 'Long', 'NIRCam channel: long or short')
        # filter, pupil added by calc_psf header code
        hdulist[0].header['PILIN'] = ('False', 'Pupil imaging lens in optical path: T/F')
        # Update APERNAME if coronagraphy
        if (self._image_mask is not None) and (self._image_mask in self._image_mask_apertures):
            result[0].header["APERNAME"] = (self._image_mask_apertures[self._image_mask], "SIAF aperture name")
```

This is only a partial fix as I don't see a simple way of doing the same procedure for Lyto stops without an occulting mask, or deciding on full frame apertures versus subarrays.

Any suggestions before I write this up as a PR?

**Comment by mperrin** [comment]
Hi @JarronL - I think this all makes sense, and if you want to put in a PR that would be great. 

Full frame vs. subarrays is a good question; in general webbpsf has never yet strived to incorporate subarray coordinates in any full fidelity, leaving that for tools like pyNRC and MIRAGE. 

For Lyot stops without occulting masks we could make reasonable defaults based on whether the module is SW or LW. Something like this pseudocode: 
```
if image_mask is None and pupil_mask is Lyot:
    if new_channel is 'long': 
        default to the LWB aperture if it's the LWB Lyot, else to 335R
    else:
        default to the SWB aperture if it's the SWB Lyot, else to 210R
```
Would that be sufficient? I don't know enough of the details of which apertures are defined in the SIAF for this; I would assume there ought to be shifted versions of all of the full-frame detector apertures?

**Comment by mperrin** [comment]
Closed by PR #351
